Promote PodTopologySpread to stable in 1.19

Huang-Wei opened the pull request:
/sig scheduling

savitharaghunathan commented:
/milestone 1.19

Huang-Wei commented:
@sftim Thanks for the reminder. This PR is ready for review.

I've rebased with dev-1.19. IIRC it's not necessary to manually update the featuregate in `content/en/docs/reference/command-line-tools-reference/{component}.md`, as there is an automation script doing that, right? If so, in this PR, I just updated `content/en/docs/reference/command-line-tools-reference/feature-gates.md` manually.

sftim commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -100,7 +100,7 @@ different Kubernetes components.
 | `ExperimentalHostUserNamespaceDefaulting` | `false` | Beta | 1.5 | |
 | `EvenPodsSpread` | `false` | Alpha | 1.16 | 1.17 |
-| `EvenPodsSpread` | `true` | Beta | 1.18 | |
+| `EvenPodsSpread` | `true` | Beta | 1.18 | 1.18 |
```
I'm afraid you need to do this differently. Remove the feature gate from _Feature gates for Alpha or Beta features_ and move it to the table for stable and graduated features.

Huang-Wei replied:
Ah, definitely. Updated.

celestehorgan commented:
/label needs-rebase

@Huang-Wei could you rebase this please?

Huang-Wei commented:
@celestehorgan sure, rebase completed.

savitharaghunathan commented:
/assign

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -265,13 +256,11 @@ scheduled - more packed or more scattered.
   single topology domain.
 
-The "EvenPodsSpread" feature provides flexible options to distribute Pods evenly across different
+The "PodTopologySpread" feature provides flexible options to distribute Pods evenly across different
```
You don't need to name feature gates for graduated features.

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -265,13 +256,11 @@ scheduled - more packed or more scattered.
 ## Known Limitations
 
 - Scaling down a Deployment may result in imbalanced Pods distribution.
 - Pods matched on tainted nodes are respected. See [Issue 80921](https://github.com/kubernetes/kubernetes/issues/80921)
```
Is this line still relevant?

Huang-Wei replied:
Yes.

sftim commented:
@kubernetes/sig-scheduling-pr-reviews please take a look - if happy, add /lgtm

Huang-Wei commented:
@sftim Thanks for the review. Comments addressed. PTAL.

annajung commented:
Hi 👋 1.19 docs shadow here,
Can we get a tech review from SIG Scheduling? @kubernetes/sig-scheduling-pr-reviews

Huang-Wei commented:
@alculquicondor Can you please take another look, if possible /lgtm?

alculquicondor commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -265,13 +256,14 @@ scheduled - more packed or more scattered.
-The "EvenPodsSpread" feature provides flexible options to distribute Pods evenly across different
-topology domains - to achieve high availability or cost-saving. This can also help on rolling update
-workloads and scaling out replicas smoothly. See [Motivation](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/895-pod-topology-spread#motivation) for more details.
+To fill the gap, you can specify topology spread constraints to distribute Pods
+across different topology domains flexibly - to achieve either high availability
```
```suggestion
For finer control, you can specify topology spread constraints to distribute Pods
across different topology domains - to achieve either high availability
```

alculquicondor commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -224,6 +222,9 @@ different Kubernetes components.
 | `ExperimentalCriticalPodAnnotation` | `false` | Deprecated | 1.13 | - |
+| `EvenPodsSpread` | `false` | Alpha | 1.16 | 1.17 |
+| `EvenPodsSpread` | `true` | Beta | 1.18 | 1.18 |
+| `EvenPodsSpread` | `true` | GA | 1.19 | - |
```
Isn't the gate removed in 1.19?

Huang-Wei replied:
I think it serves as listing historical changes.

alculquicondor replied:
@sftim for authoritative answer, but the historical purpose should be satisfied with the Alpha and Beta rows.

kbhawkey replied:
This looks correct. Once a feature reaches GA, move the alpha+beta items to `stable/deprecated` table and add `GA or deprecated`.

savitharaghunathan commented:
Hi @Huang-Wei 👋 , Please rebase to fix the merge conflict.

Huang-Wei commented:
Github seems to have a problem in showing up the latest code changes of a branch. For instance, my [forked repo](https://github.com/Huang-Wei/website/commits/1.19-pts-stable) has the latest change of this PR, but not showing up here.

Huang-Wei commented:
> Almost forgot, could you update the interpretation of whenUnschedulable and maxSkwe as we updated it in the API docs.

@alculquicondor PTAL. Also I added a link to the blog in the end.

Huang-Wei commented:
Kindly ping @sftim for a final look.

kbhawkey commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -16,13 +14,6 @@ You can use _topology spread constraints_ to control how {{< glossary_tooltip te
 
 ## Prerequisites
```
@Huang-Wei ,
Are there prerequisites?

kbhawkey replied:
Could you rename this heading?

kbhawkey commented:
Previews:
https://deploy-preview-21115--kubernetes-io-vnext-staging.netlify.app/docs/reference/command-line-tools-reference/feature-gates/

kbhawkey commented:
@Huang-Wei , looks fine. Just questioned the `Prerequisites` heading.
/lgtm

savitharaghunathan commented:
@kbhawkey Is it good to merge or the question regarding Prerequisites needs to be addressed before merging?

kbhawkey commented:
@alculquicondor , Would you provide a technical LGTM? Thanks!

kbhawkey commented:
It can merge as is.

alculquicondor commented:
I would put the feature gate as deprecated, given that it was removed.

On Wed., Jul. 15, 2020, 8:52 p.m. Karen Bradshaw, <notifications@github.com>
wrote:

> It can merge as is.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/website/pull/21115#issuecomment-659090390>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJ5E6CKSOMBIV3DSICGTA3R3ZFLBANCNFSM4NHHSU7Q>
> .
>

kbhawkey commented:
> I would put the feature gate as deprecated, given that it was removed.

Checking: EvenPodsSpread is the feature that has reached GA in 1.19?
A feature reaches GA, not the feature gate.
```
If a feature is stable you can find all stages for that feature listed in the
  [Graduated/Deprecated feature gate table](#feature-gates-for-graduated-or-deprecated-features).
```
It is somewhat confusing as the blog post mentions `PodTopologySpread`.

alculquicondor commented:
Oh I see. I was misinterpreting the role of this list.

We renamed the underlying scheduling plugin name. But the feature name
remained as originally.

/lgtm

On Wed., Jul. 15, 2020, 10:21 p.m. Karen Bradshaw, <notifications@github.com>
wrote:

> I would put the feature gate as deprecated, given that it was removed.
>
> Checking: EvenPodsSpread is the feature that has reached GA in 1.19?
> A feature reaches GA, not the feature gate.
>
> If a feature is stable you can find all stages for that feature listed in the
>   [Graduated/Deprecated feature gate table](#feature-gates-for-graduated-or-deprecated-features).
>
> It is somewhat confusing as the blog post mentions PodTopologySpread.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/website/pull/21115#issuecomment-659117084>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJ5E6EWMQMMUNDMFWGNMLLR3ZP4FANCNFSM4NHHSU7Q>
> .
>

Huang-Wei commented:
@kbhawkey The feature gate was named `EvenPodsSpread` since alpha, however, we realized that technically the word "PodTopologySpread" is more proper.

kbhawkey commented:
Great! @savitharaghunathan , this is ready to merge. Thanks.

sftim commented on content/en/docs/concepts/workloads/pods/pod-topology-spread-constraints.md:
```diff
@@ -70,7 +61,15 @@ spec:
+  It's the maximum permitted difference between the number of matching Pods in
+  any two topology domains of a given topology type. It must be greater than
+  zero. Its semantics differs according to the value of `whenUnsatisfiable`:
+  - when `whenUnsatisfiable` equals to "DoNotSchedule", `maxSkew` is the maximum
+    permitted difference between the number of matching pods in the target
+    topology and the global minimum.
+  - when `whenUnsatisfiable` equals to "ScheduleAnyway", scheduler gives higher
```
_nit (OK to skip / fix after merge):_
“is equal to” or “equals”, not “equals to” (that last phrasing is not idiomatic English)

kbhawkey commented:
/approve
